File: problems/010-all-or-nothing-output-collection/solutions/javascript/main.js
import * as std from "std";
/** @typedef {{ p: string, m: bigint, a: bigint }} File */
/** @type {string} */
const input = std.in.readAsString();
let cursor = 0;
/** @returns {string} */
function nextToken() {
  while (cursor < input.length && input.charCodeAt(cursor) <= 32) cursor++;
  const start = cursor;
  while (cursor < input.length && input.charCodeAt(cursor) > 32) cursor++;
  return input.slice(start, cursor);
}
const N = Number(nextToken()),
  Q = Number(nextToken()),
  U = BigInt(nextToken()),
  a = /** @type {File[]} */ ([]);
for (let i = 0; i < N; i++) {
  a.push({ p: nextToken(), m: BigInt(nextToken()), a: BigInt(nextToken()) });
}
a.sort((x, y) => x.p < y.p ? -1 : x.p > y.p ? 1 : 0);
const pre = /** @type {bigint[]} */ ([0n]);
let mismatch = N;
for (let i = 0; i < N; i++) {
  pre.push(pre[i] + a[i].m);
  if (mismatch === N && a[i].m !== a[i].a) mismatch = i;
}
let k = 0;
let output = "";
/** @param {string} line */
function emit(line) {
  if (output.length + line.length > 65536) {
    std.out.puts(output);
    output = "";
  }
  output += line;
}
for (let z = 0; z < Q; z++) {
  const b = BigInt(nextToken());
  if (b < U) {
    emit("ERR QUOTA -\n");
    continue;
  }
  const cap = b - U;
  while (k < N && pre[k + 1] <= cap) k++;
  if (k < mismatch) emit(`ERR QUOTA ${a[k].p}\n`);
  else if (mismatch < N) emit(`ERR MISMATCH ${a[mismatch].p}\n`);
  else if (k < N) emit(`ERR QUOTA ${a[k].p}\n`);
  else emit(`OK ${N} ${U + pre[N]}\n`);
}
if (output.length) std.out.puts(output);

File: problems/010-all-or-nothing-output-collection/solutions/cpp/main.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>
using namespace std;
struct File {
  string p;
  uint64_t m, a;
};
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  uint64_t u;
  if (!(cin >> n >> q >> u))
    return 0;
  vector<File> f(n);
  for (auto &x : f)
    cin >> x.p >> x.m >> x.a;
  sort(f.begin(), f.end(),
       [](const File &a, const File &b) { return a.p < b.p; });
  vector<uint64_t> pre(n + 1);
  int mismatch = n;
  for (int i = 0; i < n; i++) {
    pre[i + 1] = pre[i] + f[i].m;
    if (mismatch == n && f[i].m != f[i].a)
      mismatch = i;
  }
  int k = 0;
  while (q--) {
    uint64_t b;
    cin >> b;
    if (b < u) {
      cout << "ERR QUOTA -\n";
      continue;
    }
    uint64_t cap = b - u;
    while (k < n && pre[k + 1] <= cap)
      k++;
    if (k < mismatch)
      cout << "ERR QUOTA " << f[k].p << '\n';
    else if (mismatch < n)
      cout << "ERR MISMATCH " << f[mismatch].p << '\n';
    else if (k < n)
      cout << "ERR QUOTA " << f[k].p << '\n';
    else
      cout << "OK " << n << ' ' << u + pre[n] << '\n';
  }
}
